googletest/include/gtest/gtest-printers.h

```diff
@@ -677,30 +672,107 @@ void PrintTo(const ::std::pair<T1, T2>& value, ::std::ostream* os) {
   *os << ')';
 }
 
+namespace universal_printer_impl {
+
+namespace unconstrained_print_to {
+// Create a PrintTo which matches anything, and returns a type that
+// no other PrintTo overload can return
+struct SentinelType {};
+template <class T>
+SentinelType PrintTo(const T&, ::std::ostream* os);
+}  // namespace unconstrained_print_to
+
+namespace print_to_with_unconstrained_print {
+// Bring in all the PrintTo overloads in the same namespace
+// so they have the same priority (together with the ADL one)
+using internal::PrintTo;
+using unconstrained_print_to::PrintTo;
+
+// Detect the returned type of calling PrintTo in the current context.
+// It can result in 3 outcomes:
+//  1. There is a better match than the unconstrained PrintTo.
+//     It will evalute to whatever that returns, likely `void`.
+//  2. There isn't a better match than the unconstrained PrintTo we defined.
+//     It will evaluate to `SentinelType`.
+//  3. The call is ambiguous as since there are multiple valid overloads.
+//     This will trigger a template substitution failure
+template <class T>
+using PrintToOverloadResult = decltype(PrintTo(
+    ::std::declval<const T&>(), ::std::declval<::std::ostream*>()));
+
+}  // namespace print_to_with_unconstrained_print
+
+// This is going to be:
+//  1. True: when the unconstrained PrintTo would be the preferred one, if it
+//  existed
+//  2. False: when there is a PrintTo that is preferred over the unconstrained
+//  one, or if the call is ambiguous
+template <class T>
```

Comment:
There's a lot going on here that I think can be simplified.

Rather than using the sentinel and the detector idiom, would it work to

```
template <typename T>
auto CallPrintTo(T&& t, std::ostream* os, int) -> decltype(PrintTo(std::forward<T>(t), os)) {
  return PrintTo(std::forward<T>(t), os);
}
template <typename T>
void CallPrintTo(const T& t, std::ostream* os, char) {
  PrintToFallback(t, os);
}
```
And then call `CallPrintTo(t, os, 0);`?

I might be missing something here entirely, but I believe that would remove a lot of the machinery.

Reply:
I think what you are proposing is similar to #3682. Is that correct?

The problem is implicit conversions. e.g. `int` implicitly converts to `char`, so even if `PrintTo` is only defined for `char`, this will detect it as valid for `int`, and will call that instead of the fallback (which instead is the current behaviour).

If the problem was only for standard types, we could fix it in gtest side by providing all the possible overloads, but the problem exists for any class which is implicitly convertible to another one (and especially derived to base conversions).

This is the behaviour currently implemented in gtest: https://godbolt.org/z/MYzz6foEo
This is with the code you proposed: https://godbolt.org/z/3849svGx6

If we want to maintain backward compatibility, we'd need to check if `PrintTo` is invokable without implicit conversions, and the current (complicated, I agree), machinery is the only way I found.

Do you have some idea of how we could do that check in a simpler way?

Reply:
Any opinion on the above?

The alternative is to offer an extra customization point, which gets called with precedence and allows the user to define a SFINAE template?
I didn't initially propose that because I didn't want to complicate the interface for the user, but if that's preferable we can go down that route